pennylane/capture/capture_transforms.py

```diff
@@ -0,0 +1,348 @@
+        a Catalyst variant jaxpr.
+        """
+        self._trace = None
+        # _state has separate dictionaries to store global states for each transform
+        self._state = [{} for _ in range(len(self._transform_program) + 1)]
+
+    def read_with_trace(self, var):
+        """Extract the value corresponding to a variable and box it in a ``TransformTracer``
+        if not already boxed."""
+        if getattr(var, "_trace", None) is self._trace:
+            return var
+
+        return self._trace.full_raise(var)
+
+    def interpret_operation_eqn(self, eqn: "jax.core.JaxprEqn"):
+        """Interpret an equation corresponding to an operator.
+
+        Args:
+            eqn (jax.core.JaxprEqn): a jax equation for an operator.
+
+        See also: :meth:`~.interpret_operation`.
+
+        """
+        invals = [self.read(invar) for invar in eqn.invars]
+        traced_invals = [self.read_with_trace(inval) for inval in invals]
+        if isinstance(eqn.outvars[0], jax.core.DropVar):
+            return eqn.primitive.bind(*traced_invals, **eqn.params)
+
+        with qml.QueuingManager.stop_recording():
+            op = eqn.primitive.impl(*invals, **eqn.params)
+
+        # This is used for binding operators that are consumed by measurements. By saving the
+        # operator to the environment, we can use it when interpreting measurement primitives
+        # to transform observables differently than gates.
+        self._env[id(op)] = (traced_invals, eqn.params)
+        return op
+
+    def interpret_measurement_eqn(self, eqn: "jax.core.JaxprEqn"):
+        """Interpret an equation corresponding to a measurement process.
+
+        Args:
+            eqn (jax.core.JaxprEqn)
+
+        See also :meth:`~.interpret_measurement`.
+
+        """
+        invals = [self.read(invar) for invar in eqn.invars]
+
+        traced_invals = []
+        for inval in invals:
+            # The following branch is added because we want observables to get transformed.
+            # However, due to the logic used in `interpret_operator_eqn`, we only transform
+            # operators that do not get consumed by other primitives. So, we do the transforming
+            # (binding) here instead. The global state is updated with the "op_is_observable" key
+            # because transforms may want special handling for observables of measurements.
+            if isinstance(inval, qml.operation.Operator):
+                # pylint: disable=protected-access
+                op_tracers, op_params = self._env[id(inval)]
+                self._state[-1]["is_measurement_obs"] = True
+                new_inval = inval._primitive.bind(*op_tracers, **op_params)
+                self._state[-1].pop("is_measurement_obs")
+                traced_invals.append(self.read_with_trace(new_inval))
+                continue
+            traced_invals.append(self.read_with_trace(inval))
```

Comment:
This is new. I hope that the comment I've left to explain why it was added is enough. Please let me know if you need more info.